File: clientserver/src/mill/clientserver/Locks.scala
package mill.clientserver

import java.io.RandomAccessFile
import java.nio.channels.FileChannel
import java.util.concurrent.locks.{ReadWriteLock, ReentrantLock}


trait Lock{
  def lock(): Locked
  def lockBlock[T](t: => T): T = {
    val l = lock()
    try t
    finally l.release()
  }
  def tryLockBlock[T](t: => T): Option[T] = {
    tryLock() match{
      case None =>
        None
      case Some(l) =>
        try Some(t)
        finally l.release()
    }

  }
  def tryLock(): Option[Locked]
  def await(): Unit = {
    val l = lock()
    l.release()
  }

  /**
    * Returns `true` if the lock is *available for taking*
    */
  def probe(): Boolean
}
trait Locked{
  def release(): Unit
}
trait Locks{
  val processLock: Lock
  val serverLock: Lock
  val clientLock: Lock
}
class FileLocked(lock: java.nio.channels.FileLock) extends Locked{
  def release() = {
    lock.release()
  }
}

class FileLock(path: String) extends Lock{

  val raf = new RandomAccessFile(path, "rw")
  val chan = raf.getChannel
  def lock() = {
    val lock = chan.lock()
    new FileLocked(lock)
  }
  def tryLock() = {
    chan.tryLock() match{
      case null => None
      case lock => Some(new FileLocked(lock))
    }
  }
  def probe(): Boolean = tryLock() match{
    case None => false
    case Some(locked) =>
      locked.release()
      true
  }
}
class FileLocks(lockBase: String) extends Locks{
  val processLock = new FileLock(lockBase + "/pid")

  val serverLock = new FileLock(lockBase + "/serverLock")

  val clientLock = new FileLock(lockBase + "/clientLock")
}
class MemoryLocked(l: java.util.concurrent.locks.Lock) extends Locked{
  def release() = l.unlock()
}

class MemoryLock() extends Lock{
  val innerLock = new ReentrantLock(true)

  def probe() = !innerLock.isLocked
  def lock() = {
    innerLock.lock()
    new MemoryLocked(innerLock)
  }
  def tryLock() = {
    innerLock.tryLock() match{
      case false => None
      case true => Some(new MemoryLocked(innerLock))
    }
  }
}
class MemoryLocks() extends Locks{
  val processLock = new MemoryLock()

  val serverLock = new MemoryLock()

  val clientLock = new MemoryLock()
}

File: clientserver/src/mill/clientserver/Locks.java
package mill.clientserver;

import java.io.RandomAccessFile;
import java.nio.channels.FileChannel;
import java.util.concurrent.locks.ReentrantLock;


abstract class Lock{
    abstract public Locked lock() throws Exception;
    abstract public Locked tryLock() throws Exception;
    public void await() throws Exception{
        lock().release();
    }

    /**
      * Returns `true` if the lock is *available for taking*
      */
    abstract public boolean probe() throws Exception;
}
interface Locked{
    void release() throws Exception;
}
class Locks{
    public Lock processLock;
    public Lock serverLock;
    public Lock clientLock;
    static Locks files(String lockBase) throws Exception{
        return new Locks(){{
            processLock = new FileLock(lockBase + "/pid");

            serverLock = new FileLock(lockBase + "/serverLock");

            clientLock = new FileLock(lockBase + "/clientLock");
        }};
    }
    static Locks memory(){
        return new Locks(){{
            this.processLock = new MemoryLock();
            this.serverLock = new MemoryLock();
            this.clientLock = new MemoryLock();
        }};
    }
}
class FileLocked implements Locked{
    private java.nio.channels.FileLock lock;
    public FileLocked(java.nio.channels.FileLock lock){
        this.lock = lock;
    }
    public void release() throws Exception{
        this.lock.release();
    }
}

class FileLock extends Lock{
    String path;
    RandomAccessFile raf;
    FileChannel chan;
    public FileLock(String path) throws Exception{
        this.path = path;
        raf = new RandomAccessFile(path, "rw");
        chan = raf.getChannel();
    }

    public Locked lock() throws Exception{
        return new FileLocked(chan.lock());
    }
    public Locked tryLock() throws Exception{
        java.nio.channels.FileLock l = chan.tryLock();
        if (l == null) return null;
        else return new FileLocked(l);
    }
    public boolean probe()throws Exception{
        java.nio.channels.FileLock l = chan.tryLock();
        if (l == null) return false;
        else {
            l.release();
            return true;
        }
    }
}
class MemoryLocked implements Locked{
    java.util.concurrent.locks.Lock l;
    public MemoryLocked(java.util.concurrent.locks.Lock l){
        this.l = l;
    }
    public void release() throws Exception{
        l.unlock();
    }
}

class MemoryLock extends Lock{
    ReentrantLock innerLock = new ReentrantLock(true);

    public boolean probe(){
          return !innerLock.isLocked();
    }
    public Locked lock() {
        innerLock.lock();
        return new MemoryLocked(innerLock);
    }
    public Locked tryLock() {
        if (innerLock.tryLock()) return new MemoryLocked(innerLock);
        else return null;
    }
}
